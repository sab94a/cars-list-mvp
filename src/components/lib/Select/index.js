// @flow

import React from 'react';
import cx from 'classnames';
import { SelectItem, Select as SelectType } from 'components/lib/types'
import styles from './index.module.scss';

export default class Select extends React.PureComponent<SelectType> {
    static placeholder = 'None';
    static clearSelectValue = '/';

    static defaultProps = {
        defaultValue: null,
        placeholder: Select.placeholder,
        options: [],
        onChange: () => null
    };

    state = {
        active: null,
        open: false
    }

    handlers = {}

    get emptyOption(){ 
        return {
            title: this.props.placeholder,
            value: Select.clearSelectValue
        }
    }

    get items() {
        return [this.emptyOption, ...this.props.options]
    };

    componentDidMount() {
        const { defaultValue } = this.props;

        if (defaultValue) {
            this.setState({
                active: this.findItem(defaultValue).title
            })
        }
    }

    findItem = (value:string) => {
        for (let item of this.props.options) {
            if (item.value === value) {
                return item;
            };
        }
        return this.emptyOption
    };

    bindClick(value: string) {
        if(!this.handlers[value]) {
            this.handlers[value] = (event: SyntheticEvent<*>) => {
                event.preventDefault();

                if (value === Select.clearSelectValue) {
                    this.props.onChange(false)
                } else {
                    this.props.onChange(value)
                }

                this.setState({
                    active: value,
                    open: false
                })
            }
        }
        return this.handlers[value]
    }

    togglePanel = (event: SyntheticEvent<*>) => {
        event.preventDefault();

        this.setState({ open: !this.state.open })
    }


    onBlur = () => {
        this.setState({ open: false })
    }

    render() {
        const { label, className } = this.props;
        const { open, active } = this.state;
        const selectedClasses = cx(styles.selected, { [styles.open]: open})

        return (
            <div className={ cx(className, styles.root) } tabIndex="0" onBlur={ this.onBlur }>
                { !!label && (
                    <div className={ styles.label }>{ label }</div> 
                )}
                <div className={ selectedClasses } onClick={ this.togglePanel }>
                    { this.findItem(active).title }
                </div>
                { open && (
                    <ul className={ styles.list}>
                        { this.items.map(({ value, title }: SelectItem) => (
                            <li 
                                key={ value } 
                                className={ styles.item }
                                onClick={ this.bindClick(value) }
                            >{ title }</li>
                        )) }
                    </ul> 
                )}
            </div>
        )
    }
}
