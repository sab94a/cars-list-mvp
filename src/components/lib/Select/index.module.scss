@import 'components/lib/variables.scss';

.root {
    position: relative;
    font-size: $font-size-sm;
    outline: none;
    width: $button-width * 2;
    text-align: left;
}

.list {
    position: absolute;
    top: 100%;
    left: 0;
    padding: 0;
    margin: $offset-sm 0 0 0;
    min-width: 100%;
    border-radius: $border-width-sm;
    border: $border-width-sm solid $light;
    z-index: 1;
}

.selected {
    position: relative;
    border: $border-width-sm solid $light;

    &:after {
        content: '';
        display: block;
        position: absolute;
        right: $offset-sm;
        top: 50%;
        margin-top: -$offset-sm / 2;
        width: 0;
        border: 7px solid transparent;	
        border-top: 10px solid $light;
    }

    &.open:after {
        border: 7px solid transparent;	
        margin-top: -13px;
        border-bottom: 10px solid $light;
    }
}

.selected, .item {
    display: block;
    padding: $offset-sm;
    border-radius: $border-width-sm;
    background: #fff;
    user-select: none;
    cursor: pointer;
}

.label {
    font-size: $font-size-sx;
    margin-bottom: $offset-sm;
}

.item {
    border-bottom: $border-width-sm solid $light;

    &:last-of-type {
        border-bottom: 0;
    }

    &:hover, &:focus {
        background: $primary;
        color: #fff;
    }
}
