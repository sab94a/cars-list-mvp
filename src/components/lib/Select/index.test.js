import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Select from './';

describe('<Select />', () => {
    const value = 'value_1';
    const options = [{
        value: 'value_1',
        title: 'titile_1'
    }, {
        value: 'value_2',
        title: 'title_2'
    }];
    const placeholder = 'Placeholder';
    const label = 'Label';
    const className = 'extraClass';

    let onChange;
    let component;
    let defaultProps;

    beforeEach(() => {
        onChange = jest.fn();
        defaultProps = {
            value,
            options,
            placeholder,
            onChange,
            label,
            className
        };
    });

    const create = params => {
        component = mount(<Select { ...params } />);
    };

    it('Should render correctly', () => {
        create(defaultProps);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should render label if it passed', () => {
        const $label = component.find('.label');

        expect($label).toHaveLength(1);
        expect($label.text()).toBe(label);
    });

    it('Should NOT render label if it isn\'t passed', () => {
        create({ ...defaultProps, label: null });

        const $label = component.find('.label');
        expect($label).toHaveLength(0);
    });

    it('Should add extra ClassName to root container if it passed', () => {
        const $root = component.find('.root');

        expect($root.hasClass(className)).toBeTruthy();
    });

    it('If correct value passed, should render it\'s titile', () => {
        const option = options[0];

        create({ ...defaultProps, value: option.value });

        const $selected = component.find('.selected');

        expect($selected.text()).toEqual(option.title);
    });

    it('If it was passed incorrect value, should render placeholder', () => {
        create({ ...defaultProps, value: 'any_value' });
        const $selected = component.find('.selected');

        expect($selected.text()).toEqual(placeholder);
    });

    it('Should NOT render list of options by default', () => {
        create(defaultProps);

        expect(component.find('ul')).toHaveLength(0);
    });

    it('Should render list of options by select click', () => {
        create(defaultProps);
        const $selected = component.find('.selected');

        $selected.simulate('click');
        component.update();

        expect(component.find('ul')).toHaveLength(1);
    });

    it('Should render correct default option', () => {
        const $default = component.find('li').at(0);

        expect($default.text()).toBe(placeholder);
    });

    it('Should render correct options', () => {
        const $items = component.find('li');

        expect($items).toHaveLength(options.length + 1);

        for (let i = 1; i < $items.length; i++) {
            expect($items.at(i).text()).toBe(options[i - 1].title);
        };
    });

    it('Should call onChange callback with option value on click', () => {
        create(defaultProps);
        
        component.setState({ open: true });

        const index = 1;
        const $item = component.find('li').at(index);

        $item.simulate('click');

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith(options[index - 1].value);
    });

    it('Should call onChange callback with `null` on default option click', () => {
        create(defaultProps);
        
        component.setState({ open: true });

        const $item = component.find('li').at(0);

        $item.simulate('click');

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith(null);
    });

    it('Should close list of options after click', () => {
        expect(component.find('ul')).toHaveLength(0);
    });

    it('Should close list on selected click, if it is opened', () => {
        create(defaultProps);

        component.setState({ open: true });

        const $selected = component.find('.selected');

        $selected.simulate('click');
        component.update();

        expect(component.find('ul')).toHaveLength(0);
    });
});
