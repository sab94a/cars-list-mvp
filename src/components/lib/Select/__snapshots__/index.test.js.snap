// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Should render correctly 1`] = `
<Select
  className="extraClass"
  label="Label"
  onChange={[MockFunction]}
  options={
    Array [
      Object {
        "title": "titile_1",
        "value": "value_1",
      },
      Object {
        "title": "title_2",
        "value": "value_2",
      },
    ]
  }
  placeholder="Placeholder"
  value="value_1"
>
  <div
    className="extraClass root"
    onBlur={[Function]}
    tabIndex="0"
  >
    <div
      className="label"
    >
      Label
    </div>
    <div
      className="selected"
      onClick={[Function]}
    >
      titile_1
    </div>
  </div>
</Select>
`;
